common/src/api/internal/nexus.rs

```diff
@@ -84,13 +84,34 @@ pub struct SledInstanceState {
 
 // Oximeter producer/collector objects.
 
+/// The kind of metric producer this is.
+#[derive(Clone, Copy, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ProducerKind {
+    /// The producer is a sled-agent.
+    SledAgent,
+    /// The producer is an Oxide-managed service.
+    Service,
+    /// The producer is a Propolis VMM managing a guest instance.
+    Instance,
+}
+
 /// Information announced by a metric server, used so that clients can contact it and collect
 /// available metric data from it.
 #[derive(Clone, Debug, Deserialize, JsonSchema, Serialize, PartialEq)]
 pub struct ProducerEndpoint {
+    /// A unique ID for this producer.
     pub id: Uuid,
+    /// The kind of producer.
+    pub kind: Option<ProducerKind>,
```

Comment:
Should this be required?  Is it just that we're worried about API compatibility?

Reply:
Yeah, I was working under the assumption that this needed to be a multi-step process: introduce an optional field in the API and DB; update Propolis and Dendrite to send that field; and then make the field non-optional. If it is required, I think we have a commit which can't possibly work. Maybe that's OK?

Reply:
Got it -- this is because of the consumers outside of Omicron that are then used in other parts of Omicron.  I hadn't realized (or forgot) there was going to be a follow-on PR.  I think it might be worth writing this plan down in the PR description so it's clear what's going to happen and also what the state is at each point.  (For example, what actually happens if you were to build a TUF repo after landing this PR but before the others?)

My suggestion for stuff like this (and you may have been planning to do this already) is to get all the PRs lined up and not land any of them until we're happy with all of them and the sequence itself.  I sometimes find a problem when I get to the later ones that necessitates fixing an "earlier" one.

Reply:
> this is because of the consumers outside of Omicron

Yeah, exactly.

> I think it might be worth writing this plan down in the PR description

Sounds will, I'll update that now

> get all the PRs lined up

Yep, I've got branches in the Dendrite and Propolis repositories which add the new optional field to their registration requests.